feat: allow vfs implementations to specify m/b/a-times in file metadata

kartonrad opened the pull request:
This small change adds the file modification birth and access times to the VfsMetadata struct.

These times are encapsulated in `Option<>`, as they are not guaranteed to be available even in `std::fs::metadata`.

I have implemented the metadata timestamps for PhysicalFS, though it may be possible to support them for the in memory database as well.

### Motivation
I am planning on writing a minimal vfs implementation for MTS (Media Transfer Protocol).
I don't know yet if this is going to work, especially since there are few crates for MTS - but maybe i can make it work.

In the final product i will need access to the mtime field.

kartonrad commented:
Nooo i forgot to run cargo fmt
Maybe tomorrow

manuel-woelker commented:
Hi, thanks for the PR! Definitely useful!
I think adding some unit tests for the time fields is probably a good idea, even though the implementation is fairly simple.
Having support for timestamp information in MemoryFS would be quite neat as well!

kartonrad commented:
> I think adding some unit tests for the time fields is probably a good idea, even though the implementation is fairly simple. 

Good Idea! I don't know if we should necessarily expect ModifyTimes to update when we write to a file - maybe that isnt guaranteed on all operating systems...
Do you have an idea of how to test that?

> Having support for timestamp information in MemoryFS would be quite neat as well!

I just locked into it, and realized the trait definition immutably borrows the Filesystem
I could use a ref-cel i guess - what do you think

manuel-woelker commented:
Build looks good now 👍 

Yeah, a lot of systems also mount with `noatime` for performance reasons.
I think the best approach might be to set the times manually via the "normal" `std::fs` mechanisms (on systems that support it) and then verify that the VFS abstraction returns the correct values.

It might make sense to have a way to modify these times as well, that would certainly make testing easier.

Regarding the MemoryFS implementation, I think this should work without refcell. `AsyncMemoryFile` would need those additional fields to track the values, and then copy them out in `metadata()`. But it also might make sense to tackle that in a separate PR.

kartonrad commented:
The diff has grown 😯

I've added your suggestions to the non-async version
Namely:
- The optional set_creation/modification/access_time functions (path.rs and filesystem.rs)
  - [ ] should these remain optional? or do we force downstream libraries to think about this issue
- Implementations for physical and memory, as well as redirections in altroot and overlay fs
  - [ ] i used the [std::fs::File::set_times()](https://doc.rust-lang.org/stable/std/fs/struct.File.html#method.set_times) Method, that was stabilized in [Rust Version 1.75.0](https://releases.rs/docs/1.75.0/#stabilized-apis)
  do we update the 4th pipeline (1.63.0), or do we replace my `std` call with an external crate?
 - Tests in the` test_vfs!()` macro, that only run fully if the call to `set_..._time` doesnt return `VfsErrorKind::NotSupported`
   - [ ] do these tests meet your standard?

kartonrad commented:
MSRV would now be 1.75.0
![image](https://github.com/manuel-woelker/rust-vfs/assets/56208328/24c5d8c7-59d5-4e10-9292-b78c8c1dcd01)

kartonrad commented:
i could certaily do the work to implement this in async too sometime...

kartonrad commented:
I have added the neccesary tests to async_vfs
I have added the needed methods in path.rs, and redirected the calls in overlay and altroot

However i have not been able to find a function in async_std that works like the new `std::fs::File::set_times`
And i have not implemented the behaviour for async memory.rs

So the tests right now are all passing, because they recieve VfsErrorKind::NotSupported

manuel-woelker commented:
Thanks for the updates! I will take a closer look later this week.

manuel-woelker commented on Cargo.toml:
```diff
@@ -8,6 +8,7 @@ readme = "README.md"
 keywords = ["vfs", "virtual", "filesystem", "async"]
 license = "Apache-2.0"
 edition = "2021"
+rust-version = "1.75"
```
I think we should omit the rust version here since AFAIK this would allow projects to use an older version of the rust toolchain, as long as they don't use any features (like async) requiring 1.75.

So while not officially supported, it should still work without any issues.

kartonrad replied:
Unfortunately, compilation will still fail even with no features, due to [std::fs::File::set_times()](https://doc.rust-lang.org/stable/std/fs/struct.File.html#method.set_times)

But i'll look into replacing that call with a crate that can be used on lower versions :)
And wil lthen remove the field

kartonrad replied:
I changed all the calls that rely on the new std library to a use an old battle-tested crate instead

kartonrad replied:
MSRV is back to being unaffected now!

manuel-woelker replied:
Ah great! 👍

manuel-woelker commented on src/impls/embedded.rs:
```diff
@@ -120,12 +120,18 @@ where
             return Ok(VfsMetadata {
                 file_type: VfsFileType::File,
                 len: *len,
+                modified: None,
```
I think embedded files may contain metadata

cf. https://docs.rs/rust-embed/8.3.0/rust_embed/struct.Metadata.html

kartonrad replied:
Good catch!

manuel-woelker commented on src/impls/memory.rs:
```diff
@@ -165,6 +173,9 @@ impl FileSystem for MemoryFS {
                         file_type: VfsFileType::Directory,
                         content: Default::default(),
+                        created: SystemTime::now(),
+                        modified: None,
```
To emulate standard filesystem behaviour, I think it would make sense to set mtime and atime to the same value as ctime here.

kartonrad replied:
Will do!
(though on windows, directories never have modified and accessed fields - but i gues on linux that is different, and no problem for me to change)

kartonrad replied:
I implemented the change (though directory modified and accessed times are never modified by memoryfs)

manuel-woelker replied:
Okay, looks good

manuel-woelker commented on src/impls/memory.rs:
```diff
@@ -220,9 +236,42 @@ impl FileSystem for MemoryFS {
             file_type: file.file_type,
             len: file.content.len() as u64,
+            modified: file.modified,
+            created: Some(file.created),
```
To emulate standard filesystem behaviour, I think it would make sense to set mtime and atime to the same value as ctime here.

kartonrad replied:
Wait - this is the `metadata` function
Isn't it it's job to return all three timestamps? The MemoryFile isn't modified here, it's just returned

kartonrad replied:
I still don't understand what you mean here...
I think you'd need to explain

manuel-woelker replied:
Sorry, my bad! That comment was meant to go somewhere else, please ignore ;-)

manuel-woelker commented on src/test_macros.rs:
```diff
@@ -85,6 +159,7 @@ macro_rules! test_vfs {
                 assert_eq!(metadata.len, 0);
             }
 
+            /*
```
Did this test cause any issues?

kartonrad replied:
Ah - yeah i just commented it out without explanation.

Yes - i couldn't use the exported test_vfs macro in my own crate, since `camino` is only a dev dependency of `vfs`
So it wasn't available in my crate - maybe just installing `camino` or adding it to my dev dependencies to would fix the issue withotu removing the test
I'll try it out!

kartonrad replied:
I modified cargo.toml such that camino becomes a neccesary dependency if you export the test macros

manuel-woelker replied:
That looks like a good solution 👍

manuel-woelker reviewed (changes requested):
Looking good, some small suggestions.

The merge conflicts with master only touch the imports and should be fairly simple to resolve.

If you have any questions, let me know!

kartonrad commented:
Feel free to resolve the threads that i have fixed.

manuel-woelker commented:
Merged! Thanks again for the great work, especially getting all the finer details sorted out 👍
